### PR TITLE
[SYCLomatic] Refactor the implementation of 5 cooperative groups migration with rewriter

### PR DESCRIPTION
**Pull request by zhiweij1**
APIs: sync, tiled_partition, thread_rank, size, shfl_down

Signed-off-by: Jiang, Zhiwei <zhiwei.jiang@intel.com>

### CODE REVIEW

**Review by AndyCHHuang — APPROVED** [review]
LGTM

File: clang/lib/DPCT/CallExprRewriter.cpp [hunk 1]
```diff
@@ -1247,6 +1247,31 @@ makeCallArgCreator(std::string Str) {
   return [=](const CallExpr *C) -> const StringRef { return StringRef(Str); };
 }
 
+std::function<std::string(const CallExpr *)>
+makeCallArgCreatorFromTemplateArg(unsigned Idx) {
```
Comment (ziranzha) [inline clang/lib/DPCT/CallExprRewriter.cpp]: can reuse lambda in makeTemplatedCalleeCreator

File: clang/lib/DPCT/CallExprRewriter.cpp [hunk 2]
```diff
@@ -2422,6 +2448,28 @@ class CompareArgType {
...
+
+std::function<std::string(const CallExpr *)> MemberExprBase() {
+  return [=](const CallExpr *C) -> std::string {
+    auto ME = dyn_cast<MemberExpr>(C->getCallee()->IgnoreImpCasts());
+    if (!ME)
+      return "";
+    auto Base = ME->getBase()->IgnoreImpCasts();
+    if (!Base)
+      return "";
+    auto DRE = dyn_cast<DeclRefExpr>(Base);
```
Comment (ziranzha) [inline clang/lib/DPCT/CallExprRewriter.cpp]: Expect print correct string for base expr that is not DeclRefExpr

Comment (zhiweij1) [reply clang/lib/DPCT/CallExprRewriter.cpp]: done

File: clang/lib/DPCT/CallExprRewriter.cpp (hunk 1 above)
Comment (zhiweij1) [reply clang/lib/DPCT/CallExprRewriter.cpp]: done

File: clang/lib/DPCT/ExprAnalysis.cpp [hunk 3]
```diff
@@ -752,9 +764,9 @@ void ExprAnalysis::analyzeExpr(const CXXMemberCallExpr *CMCE) {
   PP.PrintCanonicalTypes = true;
   auto BaseType = CMCE->getObjectType().getUnqualifiedType().getAsString(PP);
 
+  std::string MethodName = "";
```
Comment (zhiweij1) [inline clang/lib/DPCT/ExprAnalysis.cpp]: revert this change

Comment (zhiweij1) [inline clang/lib/DPCT/ExprAnalysis.cpp:767]: revert this change

File: clang/lib/DPCT/ASTTraversal.cpp [hunk 4]
```diff
@@ -12882,319 +12882,208 @@ void WarpFunctionsRule::runRule(const MatchFinder::MatchResult &Result) {
...
-      } else if (ArgType == "const cooperative_groups::__v1::thread_block" ||
-                 ArgType == "const cooperative_groups::__v1::coalesced_group") {
-        report(CE->getBeginLoc(), Diagnostics::BARRIER_PERFORMANCE_TUNNING,
-               true, "nd_item");
-        std::string Replacement = DpctGlobalInfo::getItem(CE) + ".barrier()";
-        emplaceTransformation(new ReplaceStmt(CE, std::move(Replacement)));
-      } else {
-        EMIT_WARNING_AND_RETURN;
+
+  auto GetBaseTypeStr = [](const CallExpr *CE) -> std::string {
```
Comment (zhiweij1) [inline clang/lib/DPCT/ASTTraversal.cpp]: we can reuse functions in CallExprRewriter.cpp/.h

Comment (zhiweij1) [inline clang/lib/DPCT/ASTTraversal.cpp:12908]: we can reuse functions in CallExprRewriter.cpp/.h

File: clang/lib/DPCT/CallExprRewriter.cpp [hunk 5]
```diff
@@ -1405,30 +1405,47 @@ makeLambdaCreator(bool IsCaptureRef,
...
+            Ret.emplace_back(List[Idx]);
           }
         }
         return Ret;
       });
 }
 
+std::function<std::string(const CallExpr *)>
+makeCallArgCreatorFromTemplateArg(unsigned Idx) {
+  return [=](const CallExpr *CE) -> std::string {
```
Comment (zhiweij1) [inline clang/lib/DPCT/CallExprRewriter.cpp:1444]:
```suggestion
  return [=](const CallExpr *CE) -> std::string {
```
```suggestion
  return [=](const CallExpr *CE) ->TemplateArgumentInfo {
```

Comment (zhiweij1) [inline clang/lib/DPCT/CallExprRewriter.cpp:1444]:
```suggestion
  return [=](const CallExpr *CE) -> std::string {
```
```suggestion
  return [=](const CallExpr *CE) ->TemplateArgumentInfo {
```

Comment (zhiweij1) [inline clang/lib/DPCT/CallExprRewriter.cpp:1444]:
```suggestion
  return [=](const CallExpr *CE) -> TemplateArgumentInfo {
```

Comment (zhiweij1) [inline clang/lib/DPCT/CallExprRewriter.cpp]:
```suggestion
  return [=](const CallExpr *CE) -> TemplateArgumentInfo {
```

Comment (zhiweij1) [reply clang/lib/DPCT/CallExprRewriter.cpp]: done

File: clang/lib/DPCT/ExprAnalysis.cpp (hunk 3 above)
Comment (zhiweij1) [reply clang/lib/DPCT/ExprAnalysis.cpp]: done

File: clang/lib/DPCT/ASTTraversal.cpp (hunk 4 above)
Comment (zhiweij1) [reply clang/lib/DPCT/ASTTraversal.cpp]: done

**Review by ziranzha — APPROVED** [review]
LGTM
